lxd/patches.go

```diff
@@ -1369,51 +1369,63 @@ func upgradeFromStorageTypeLvm(name string, d *Daemon, defaultPoolName string, d
 					// and it means that this was a
 					// mixed-storage LXD instance.
 
-					// Initialize storage interface for the new
-					// snapshot.
-					csStorage, err := storagePoolVolumeContainerLoadInit(d.State(), "default", cs)
-					if err != nil {
-						logger.Errorf("Failed to initialize new storage interface for LVM container %s: %s", cs, err)
-						return err
-					}
-
 					// Load the snapshot from the database.
 					csStruct, err := instance.LoadByProjectAndName(d.State(), "default", cs)
 					if err != nil {
 						logger.Errorf("Failed to load LVM container %s: %s", cs, err)
 						return err
 					}
 
-					// Create an empty LVM logical volume
-					// for the snapshot.
-					err = csStorage.ContainerSnapshotCreateEmpty(csStruct)
+					pool, err := storagePools.GetPoolByInstance(d.State(), csStruct)
 					if err != nil {
-						logger.Errorf("Failed to create empty LVM logical volume for container %s: %s", cs, err)
 						return err
 					}
 
-					// In case the new LVM logical volume
-					// for the snapshot is not mounted mount
-					// it.
-					if !shared.IsMountPoint(newSnapshotMntPoint) {
-						_, err = csStorage.ContainerMount(csStruct)
-						if err != nil {
-							logger.Errorf("Failed to mount new empty LVM logical volume for container %s: %s", cs, err)
-							return err
-						}
+					parent, _, _ := shared.InstanceGetParentAndSnapshotName(csStruct.Name())
+					parentInst, err := instance.LoadByProjectAndName(d.State(), csStruct.Project(), parent)
+					if err != nil {
+						logger.Errorf("Failed to load parent LVM container %s: %s", cs, err)
+						return err
 					}
 
-					// Use rsync to fill the empty volume.
-					output, err := rsync.LocalCopy(oldSnapshotMntPoint, newSnapshotMntPoint, "", true)
+					// Create an empty LVM logical volume for the snapshot.
+					err = pool.CreateInstanceSnapshot(csStruct, parentInst, nil)
```

Comment:
@stgraber I've used `CreateInstanceSnapshot` here rather than `CreateInstance`, as it seems more appropriate, however it is not an exact match to the previous use of `ContainerSnapshotCreateEmpty`, as the new storage layer doesn't have the equivalent to create an empty snapshot volume. `CreateInstance` would probably work to create an empty snapshot volume as it doesn't have an checks for snapshot instances being passed in. However I thought it would be clearer to use `CreateInstanceSnapshot` and then just have the rsync call further down wipe over it.

Reply:
Yeah, that's fine